errors/middleware-upgrade-guide.md

```diff
@@ -0,0 +1,311 @@
+| `cookie.delete`         | `clearCookie` |
+| `cookie.getWithOptions` | `cookies`     |
+
+### Explanation
+
+Based on beta feedback, we are changing the Cookies API in `NextRequest` and `NextResponse` to align more to a `get`/`set` model. The `Cookies` API extends Map, including methods like [entries](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries) and [values](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries).
+
+### How to upgrade
+
+`NextResponse` now has a `cookies` instance with:
+
+- `cookie.delete`
+- `cookie.set`
+- `cookie.getWithOptions`
+
+As well as other extended methods from `Map`.
+
+#### Before
+
+```javascript
+import { NextResponse } from 'next/server'
+import type { NextRequest } from 'next/server'
+
+export function middleware(request: NextRequest) {
+  // create an instance of the class to access the public methods. This uses `next()`,
+  // you could use `redirect()` or `rewrite()` as well
+  let response = NextResponse.next()
+  // get the cookies from the request
+  let cookieFromRequest = request.cookies['my-cookie']
+  // set the `cookie`
+  response.cookie('hello', 'world')
+  // set the `cookie` with options
+  const cookieWithOptions = response.cookie('hello', 'world', {
+    path: '/',
+    maxAge: 1000 * 60 * 60 * 24 * 7,
+    httpOnly: true,
+    sameSite: 'strict',
+    domain: 'example.com',
+  })
+  // clear the `cookie`
+  response.clearCookie('hello')
+
+  return response
+}
+```
+
+#### After
+
+```typescript
+// middleware.ts
+export function middleware() {
+  const response = new NextResponse()
+
+  // set a cookie
+  response.cookies.set('vercel', 'fast')
+
+  // set another cookie with options
+  response.cookies.set('nextjs', 'awesome', { path: '/test' })
+
+  // get all the details of a cookie
+  const [value, options] = response.cookies.getWithOptions('vercel')
+  console.log(value) // => 'fast'
+  console.log(options) // => { Path: '/test' }
+
+  // delete a cookie means mark it as expired
+  response.cookies.delete('vercel')
+
+  // clear all cookies means mark all of them as expired
+  response.cookies.clear()
+}
+```
+
+## No More Page Match Data
+
+### Summary of changes
+
+- Use [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) to check if a Middleware is being invoked for a certain page match
+
+### Explanation
+
+Currently, Middleware estimates whether you are serving an asset of a Page based on the Next.js routes manifest (internal configuration). This value is surfaced through `request.page`.
+
+To make page and asset matching more accurate, we are now using the web standard `URLPattern` API.
+
+### How to upgrade
+
+Use [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLPattern) to check if a Middleware is being invoked for a certain page match.
+
+#### Before
+
+```typescript
+// middleware.ts
+import { NextRequest, NextResponse } from 'next/server'
```

Comment:
NextRequest should be imported as a type, as a separate import. Although you can do it this way on latest TS version, for backwards compatibility we should show the "old" way and save ourselves having to explain and recommend updating TS versions